### ISSUE TITLE
Running both apache server and nginx server on same host

### ISSUE THREAD

**Issue by M-Stenzel** [yrutschle/sslh#321]
Hi team, 

this was asked before ( https://github.com/yrutschle/sslh/issues/127 ) but I was not able to solve this task myself (not a REAL expert...)

Thing is, that sometimes it is VERY convenient to run both apache and nginx on the same host (same IP).

What I want to achieve:

Address "dog.someserver.org" should redirect to port 1443 where apache is listening, and
address "cat.someserver.org" should redirect to port 2443 where nginx is listening.

How am I able to achieve this? It would be cool to have sth. like a "cooking recipe" which includes all the necessary steps to follow :)

TIA, Martin.

P. S. What would it look like if address "mouse.someserver.org" should redirect to port 2443 (nginx) as well?

**Comment by yrutschle** [comment]
This is trivial with `sni_hostnames`, as show in `example.cfg`, so you'd have something like:

```
protocols: (
{ name: "tls"; host: "localhost"; port: "1443"; sni_hostnames: [ "dog.someserver.org"  ];  },
{ name: "tls"; host: "localhost"; port: "2443"; sni_hostnames: [ "cat.someserver.org"  ];  }
);
```

I suppose I can add a 'recipe' in the FAQ. Out of curiosity, where did you look for information? Where would you have expected to find this information? Currently, `example.cfg` is the main documentation for the configuration format, but I'm not completely sure I actually say so, anywhere...

**Comment by M-Stenzel** [comment]
> This is trivial with `sni_hostnames`, as show in `example.cfg`, so you'd have something like:
> 
> ```
> protocols: (
> { name: "tls"; host: "localhost"; port: "1443"; sni_hostnames: [ "dog.someserver.org"  ];  },
> { name: "tls"; host: "localhost"; port: "2443"; sni_hostnames: [ "cat.someserver.org"  ];  }
> );
> ```
> 
> I suppose I can add a 'recipe' in the FAQ. Out of curiosity, where did you look for information? Where would you have expected to find this information? Currently, `example.cfg` is the main documentation for the configuration format, but I'm not completely sure I actually say so, anywhere...




Huh huh huh, took me some time, well this is sth. of a "recipe" - thank YOU. I was able to apply and am quite happy with things. Now I run (thanks to docker philosophy...) three webservers on one host and things look quite nice.
FYI: I looked for information on the web in general and on github.

Sadly, another problem popped up: I run openvidu ([https://openvidu.io/](https://openvidu.io/)) with docker  ([https://hub.docker.com/r/openvidu/openvidu-server](https://hub.docker.com/r/openvidu/openvidu-server)). Using the conference software from home works perfectly, however when port 4443 is blocked by a firewall, websockets fail...

Browser error message:
WebSocket connection to 'wss://openvidu.xy-space.de:4443/openvidu?sessionId=open-gold-weasel' failed: Error during WebSocket handshake: Unexpected response code: 403


Now I wonder: to bypass the firewall sslh should redirect wss://openvidu.xy-space.de:4443 to wss://openvidu.xy-space.de:443 but this does not seem to happen.

Are websockets supported? How to configure?
Does the protocol "tls" apply here?


BTW, this is part of my sslh.config

---
verbose: 0;
numeric: false;
transparent: false;
timeout: 2;
user: "sslh";

listen:
(
        { host: "0.0.0.0"; port: "443"; }
);

protocols:
(
    ...
        { name: "tls"; host: "127.0.0.1"; port: "4443"; sni_hostnames: [ "openvidu.xy-space.de" ]; log_level: 0; tfo_ok: true },
    ...

---


Martin.
